graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/method/SourceArgumentResolver.kt

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2025 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.graphql.dgs.internal.method
+
+import com.netflix.graphql.dgs.Source
+import graphql.schema.DataFetchingEnvironment
+import org.springframework.core.MethodParameter
+
+class SourceArgumentResolver : ArgumentResolver {
+    override fun supportsParameter(parameter: MethodParameter): Boolean = parameter.hasParameterAnnotation(Source::class.java)
+
+    override fun resolveArgument(
+        parameter: MethodParameter,
+        dfe: DataFetchingEnvironment,
+    ): Any {
+        val source = dfe.getSource<Any>()
+        if (source != null && parameter.parameterType == source.javaClass) {
+            return source
+        } else {
+            throw IllegalArgumentException(
+                "Invalid source type '${source?.javaClass?.name}'. Expected type '${parameter.parameterType.name}'",
```

Comment:
If this is null, what would the message be? Will it indicate that source is null?

Reply:
Can source even be null? I'm guessing graphql-java will throw an exception before it gets to this?

Reply:
I don't think it can be null because the datafetcher wouldn't even be called

Reply:
based on the interface definition:
```
@PublicApi
public interface DataFetchingEnvironment extends IntrospectionDataFetchingEnvironment {
        @Nullable
        <T> T getSource();
...
}
```
I guess it can

Reply:
Actually it could be null when using `@Source` when there is no source, e.g. on a `@DgsQuery`. Added better handling and test.